server/src/app.ts

```diff
@@ -82,17 +78,37 @@ export const main = async (app: Express) => {
           },
         },
       },
+      update: {},
     });
   }
 
+  async function findOrCreateUser(email: string) {
+    try {
+      return await upsertUser(email);
+    } catch (e) {
+      if (
+        e instanceof Prisma.PrismaClientKnownRequestError &&
+        e.code === UNIQUE_CONSTRAINT_FAILED
+      ) {
+        console.error(
+          'Received concurrent login requests - the client is likely making too many requests.',
+        );
+      } else {
+        throw e;
+      }
+    }
+  }
+
   app.post('/login', checkJwt, (req, res, next) => {
     const token = getBearerToken(req);
     if (token) {
       const userInfo = fetchUserInfo(token);
       userInfo
         .then(async ({ email }) => {
           if (!email) throw Error('No email found in user info');
-          const user = (await findUser(email)) ?? (await createUser(email));
+
+          const user = await findOrCreateUser(email);
+          if (!user) throw Error('No user found');
```

Comment:
```suggestion
```
Sorry, I hadn't realised this would be the consequence of removing the second `upsert`. I think it's a better UX if we only allow errors we can't handle to terminate the request.

With that in mind, I'll add a suggestion in `findOrCreateUser` to remove the need to handle `undefined` here.